src/vocabulary/mod.rs

```diff
@@ -0,0 +1,368 @@
+                                NormalizerWrapper::Prepend(_) => None,
+                                _ => Some(normalizer.clone()),
+                            })
+                            .collect(),
+                    );
+                    tokenizer.with_normalizer(new_sequence.into());
+                }
+                NormalizerWrapper::Prepend(_) => {
+                    tokenizer.with_normalizer(None::<NormalizerWrapper>);
+                }
+                _ => {}
+            }
+        }
+    }
+}
+
+impl TryFrom<(&mut Tokenizer, u32)> for Vocabulary {
+    type Error = VocabularyError;
+
+    fn try_from(value: (&mut Tokenizer, u32)) -> Result<Vocabulary, VocabularyError> {
+        let (tokenizer, eos_token_id) = value;
+
+        let mut vocabulary = Vocabulary::new(Some(eos_token_id));
+        for (id, added_token) in tokenizer.get_added_tokens_decoder().iter() {
+            if !added_token.special {
+                vocabulary = vocabulary.insert(added_token.content.clone(), *id);
+            }
+        }
+
+        let processor = TokenProcessor::new(tokenizer)?;
+        for (token, token_id) in tokenizer.get_vocab(false) {
+            let token_bytes = processor.process(token)?;
+            // TODO: lossy is temp:
+            // - in python in was handled by byte_symbol function
+            // - interface needs to be redefined to treat Token type as bytes: Vec<u8>
+            let processed_token = String::from_utf8_lossy(&token_bytes);
+            vocabulary = vocabulary.insert(processed_token, token_id);
+        }
+
+        Ok(vocabulary)
+    }
+}
+
+impl Vocabulary {
+    /// Inserts a token to the vocabulary with the specified identifier.
+    pub fn insert(mut self, token: impl Into<Token>, id: TokenId) -> Vocabulary {
+        self.insert_in_place(token, id);
+        self
+    }
+
+    /// Extends the vocabulary with tokens and their identifiers.
+    pub fn extend<T: Into<Token>, I: IntoIterator<Item = TokenId>>(
+        mut self,
+        tokens_and_ids: impl IntoIterator<Item = (T, I)>,
+    ) -> Vocabulary {
+        self.extend_in_place(tokens_and_ids);
+        self
+    }
+}
+
+impl Vocabulary {
+    /// Inserts a token to the vocabulary with the specified identifier, in place.
+    pub fn insert_in_place(&mut self, token: impl Into<Token>, id: TokenId) {
+        // TODO: return error if eos token id is inserted
+        let token = token.into();
+        self.map.entry(token).or_default().push(id);
+    }
+
+    /// Extends the vocabulary with tokens and their identifiers, in place.
+    pub fn extend_in_place<T: Into<Token>, I: IntoIterator<Item = TokenId>>(
+        &mut self,
+        tokens_and_ids: impl IntoIterator<Item = (T, I)>,
+    ) {
+        for (token, ids) in tokens_and_ids.into_iter() {
+            let token = token.into();
+            self.map.entry(token).or_default().extend(ids);
+        }
+    }
+}
+
+impl std::ops::Deref for Vocabulary {
```

Comment:
I'm going to need someone to ELI5 why we want/need this.
I find implementing deref to perform this kind of field selection to be morally dubious.
I, for one, expect `*` to remove a layer of pointer indirection, and that's all.